src/micro.c

```diff
@@ -170,88 +246,145 @@ microGroup_AddEndpoint(microGroup *g, microEndpointConfig *cfg)
     return micro_add_endpoint(NULL, g->m, g, cfg, false);
 }
 
-microError *
-microService_Stop(microService *m)
+static microError *
+_attach_service_to_connection(natsConnection *nc, microService *service)
 {
-    microError *err = NULL;
-    microEndpoint *ep = NULL;
-    bool finalize = false;
-    microDoneHandler doneHandler = NULL;
-
-    if (m == NULL)
+    natsStatus s = NATS_OK;
+    if (nc == NULL || service == NULL)
         return micro_ErrorInvalidArg;
 
-    _lock_service(m);
-
-    if (m->stopped)
+    natsMutex_Lock(nc->servicesMu);
+    if (nc->services == NULL)
     {
-        _unlock_service(m);
-        return NULL;
+        nc->services = NATS_CALLOC(1, sizeof(microService *));
+        if (nc->services == NULL)
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+    }
+    else
+    {
+        microService **tmp = NATS_REALLOC(nc->services, (nc->numServices + 1) * sizeof(microService *));
+        if (tmp == NULL)
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+        else
+            nc->services = tmp;
     }
-    ep = m->first_ep;
 
-    for (; ep != NULL; ep = ep->next)
+    if (s == NATS_OK)
     {
-        if (err = micro_stop_endpoint(ep), err != NULL)
-        {
-            _unlock_service(m);
-            return microError_Wrapf(err, "failed to stop service '%s', stopping endpoint '%s'", m->cfg->Name, ep->name);
-        }
+        nc->services[nc->numServices] = service;
+        nc->numServices++;
     }
+    natsMutex_Unlock(nc->servicesMu);
 
-    finalize = (m->first_ep == NULL);
-    if (finalize)
+    return micro_ErrorFromStatus(s);
+}
+
+static void
+_detach_service_from_connection(natsConnection *nc, microService *m)
+{
+    if (nc == NULL || m == NULL)
+        return;
+
+    natsMutex_Lock(nc->servicesMu);
+    for (int i = 0; i < nc->numServices; i++)
     {
-        natsLib_stopServiceCallbacks(m);
-        m->stopped = true;
-        doneHandler = m->cfg->DoneHandler;
+        if (nc->services[i] != m)
+            continue;
+        
```

Comment:
I have modified this file and when saving, I see some modifications due to spaces, etc.. You should look at VSCode `Settings/Text Editor/Trim Trailing Whitespaces` and make sure the box is checked.